blib/man3/MD5Check.3
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "MD5Check 3"
.TH MD5Check 3 "2016-11-01" "perl v5.18.4" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
MD5Check \-  Use it for init Web files's md5 values of your site(or other dir), and check if it changed!
.PP
检查web目录（或者其他重要系统目录）md5值，当目录文件变化提醒。用于文件防篡改。
.SH "VERSION"
.IX Header "VERSION"
Version 0.06
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    use MD5Check;
.Ve
.PP
## 初始化目录md5值,参数为要监控的目录
.PP
.Vb 2
\&    my $mydir=shift;
\&    print md5init($mydir);
.Ve
.PP
## 对目录文件进行检查，只需输入之前保存的md5 文件值。
.PP
.Vb 3
\&    use MD5Check;
\&    my $mydir=shift;
\&    print md5check($mydir);
\&
\&    # oneliner，perl单行程序实现功能。
.Ve
.PP
需要安装该模块，简单通过 cpanm MD5Check 安装。
.PP
.Vb 2
\&    $ perl \-MMD5Check \-e \*(Aqprint md5init("/web")\*(Aq >file
\&    $ perl \-MMD5Check \-e \*(Aqprint md5check(file) perl \-MMD5Check \-e \*(Aqmd5check(file)\*(Aq..
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
MD5Check \- Use it for init Web files's md5 values of your site(or other dir), and check if it changed.
.PP
MD5Check is not standardized. This module is far from complete.
.SS "\s-1IMPLEMENTATION\s0"
.IX Subsection "IMPLEMENTATION"
#Init the web dir md5 values
.PP
.Vb 5
\&    use MD5Check;
\&    my $mydir=shift;
\&    print  md5init($mydir);
\&=====================
\&#check files 
\&
\&    use MD5Check;
\&    my $mydir=shift;
\&    print md5check($mydir);
.Ve
.SH "Git repo"
.IX Header "Git repo"
<https://github.com/bollwarm/MD5Check.git>
.SH "AUTHOR"
.IX Header "AUTHOR"
orange \f(CW\*(C`<linzhe@ijz.me>\*(C'\fR, <http://ijz.me>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright (C) 2016 linzhe
.PP
This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.
.SH "AUTHOR"
.IX Header "AUTHOR"
orange, \f(CW\*(C`<linzhe at ijz.me>\*(C'\fR
.SH "BUGS"
.IX Header "BUGS"
Please report any bugs or feature requests to \f(CW\*(C`bug\-md5check at rt.cpan.org\*(C'\fR, or through
the web interface at <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MD5Check>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.
.SH "SUPPORT"
.IX Header "SUPPORT"
You can find documentation for this module with the perldoc command.
.PP
.Vb 1
\&    perldoc MD5Check
.Ve
.PP
You can also look for information at:
.IP "\(bu" 4
\&\s-1RT: CPAN\s0's request tracker (report bugs here)
.Sp
<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MD5Check>
.IP "\(bu" 4
AnnoCPAN: Annotated \s-1CPAN\s0 documentation
.Sp
<http://annocpan.org/dist/MD5Check>
.IP "\(bu" 4
\&\s-1CPAN\s0 Ratings
.Sp
<http://cpanratings.perl.org/d/MD5Check>
.IP "\(bu" 4
Search \s-1CPAN\s0
.Sp
<http://search.cpan.org/dist/MD5Check/>
.SH "ACKNOWLEDGEMENTS"
.IX Header "ACKNOWLEDGEMENTS"
1; # End of MD5Check
.SH "POD ERRORS"
.IX Header "POD ERRORS"
Hey! \fBThe above document had some coding errors, which are explained below:\fR
.IP "Around line 16:" 4
.IX Item "Around line 16:"
Non-ASCII character seen before =encoding in '检查web目录（或者其他重要系统目录）md5值，当目录文件变化提醒。用于文件防篡改。'. Assuming \s-1UTF\-8\s0


Changes
Revision history for MD5Check

0.06    2016-11-1/19

* change the outer of function.
* Add bin/ for the tools.

0.05    2016-10-25/20

* 增加中文文档说明

0.04    2016-10-20/20

*some changes

0.03    2016-10-08/09

*version 0.03 and for project files for cpan.

0.01    2016-09-20/

*First version, released on an unsuspecting world.



lib/MD5Check.pm
package MD5Check;

use strict;
use warnings;
use utf8;
use Digest::MD5;
require Exporter;

our @ISA    = qw(Exporter);
our @EXPORT = qw(md5init md5check);

=head1 NAME

MD5Check -  Use it for init Web files's md5 values of your site(or other dir), and check if it changed!

检查web目录（或者其他重要系统目录）md5值，当目录文件变化提醒。用于文件防篡改。

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

    use MD5Check;

## 初始化目录md5值,参数为要监控的目录

    my $mydir=shift;
    print md5init($mydir);

## 对目录文件进行检查，只需输入之前保存的md5 文件值。

    use MD5Check;
    my $mydir=shift;
    print md5check($mydir);

# oneliner，perl单行程序实现功能。

需要安装该模块，简单通过 cpanm MD5Check 安装。

## 相关实例可以直接使用bin/下的init.pl 和 check.pl 单行执行

    $ perl -MMD5Check -e 'print md5init("/web")' >file
    $ perl -MMD5Check -e 'print md5check(file) perl -MMD5Check -e 'md5check(file)'..


=cut

my $DEBUG = 0;

sub md5_sum {

    my ( $file_name, $mode ) = @_;
    my ( $FD, $ctx, $md5 );
    open( $FD, $file_name ) or die "Can't open /'$file_name/': $!";
    $ctx = Digest::MD5->new;
    binmode($FD) if $mode;
    $ctx->addfile($FD) || die "$!\n";
    $md5 = $ctx->hexdigest;
    close $FD;
    return $md5;
}

sub md5check {
    my $file = shift;
    open( my $fd, '<', $file ) or die "$file: $!\n";
    my $res .= $file . "\n";
    while (<$fd>) {
        my ( $name, $sum ) = split /\s+/;
        $name =~ s/^\*//;
        if ( $sum eq md5_sum( $name, 1 ) ) {
            $res .= "$name OK\n";
        }
        else {
            $res .= "$name FAILED\n";
        }
    }

    close $fd;

    return $res;
}

# 遍历目录计算md5值
sub md5init {

    my $fd = shift;
    my $md5value;
    my $res;
    if ( -f $fd ) {
        if ( -T $fd ) {

            #print "按照文本模式进行计算MD5!\n";
            $md5value = md5_sum( $fd, 0 );
            $res .= "$fd\t$md5value\n";
        }
        elsif ( -B $fd ) {

            #print "二进制文件用binmod计算MD5!\n";
            $md5value = md5_sum( $fd, 1 );
            $res .= "$fd\t$md5value\n";
        }
        else {
            #print "其他文件，按照bimmod计算!\n";
            $md5value = md5_sum( $fd, 1 );
            $res .= "$fd\t$md5value\n";
        }
    }
    elsif ( -d $fd ) {
        my $file_md5;
        print "开始验证目录下所有文件:\n";
        opendir( my $DH, $fd ) or die "Can't open dir $fd: $!";
        for ( readdir $DH ) {
            my $file = $fd . '/' . $_;

            # 上级目录..，本目录. 以及连接文件跳过
            next if ( $file =~ m{/.$} || $file =~ m{/..$} || -l $file );
            $res .= md5init($file);
            print "Debug::", $res if $DEBUG;
        }
        closedir $DH;
    }
    return $res;
}

1;
__END__

=head1 DESCRIPTION

MD5Check - Use it for init Web files's md5 values of your site(or other dir), and check if it changed.

MD5Check is not standardized. This module is far from complete.


=head2 IMPLEMENTATION

#Init the web dir md5 values

    use MD5Check;
    my $mydir=shift;
    print  md5init($mydir);
=====================
#check files 

    use MD5Check;
    my $mydir=shift;
    print md5check($mydir);

=head1 Git repo

L<https://github.com/bollwarm/MD5Check.git>

=head1 AUTHOR

orange C<< <linzhe@ijz.me> >>, L<http://ijz.me>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 linzhe

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut


=head1 AUTHOR

orange, C<< <linzhe at ijz.me> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-md5check at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MD5Check>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

 perldoc MD5Check


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MD5Check>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MD5Check>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MD5Check>

=item * Search CPAN

L<http://search.cpan.org/dist/MD5Check/>

=back


# End of MD5Check
